crates/ruff_linter/src/rules/pylint/rules/unnecessary_chained_comparison.rs

```diff
@@ -40,40 +40,47 @@ impl Violation for UnnecessaryChainedComparison {
     }
 }
 
-// Each integer is a unique identifier for the node.
-#[derive(Default)]
-struct Bounds {
-    lower_bound: HashSet<u32>,
-    upper_bound: HashSet<u32>,
-}
-
 fn update_bounds<'a>(
     operator: ast::CmpOp,
     id: &'a str,
     node_idx: u32,
     is_left: bool,
-    uses: &mut HashMap<&'a str, Bounds>,
+    uses: &mut HashMap<&'a str, (u32, u32)>,
 ) {
+    // Store the bounds using bitwise operations. Each bit in these
+    // integers represents whether the identifier is involved in a lower or upper bound
+    // comparison at a specific node index.
+    //
+    // For example, if node_idx is 2, then:
+    // - `1 << node_idx` produces a bitmask with the 3rd bit set.
+    // - Using |=, we set the corresponding bit in the lower or upper bounds integer to 1 without modifying other bits.
+    //
+    // This efficiently tracks whether the identifier is used in lower or upper bounds
+    // comparisons across multiple nodes, allowing us to check for shared bounds later.
     match operator {
         ast::CmpOp::Lt | ast::CmpOp::LtE if is_left => {
-            uses.entry(id).or_default().lower_bound.insert(node_idx);
+            let entry = uses.entry(id).or_default();
+            entry.0 |= 1 << node_idx;
         }
         ast::CmpOp::Gt | ast::CmpOp::GtE if is_left => {
-            uses.entry(id).or_default().upper_bound.insert(node_idx);
+            let entry = uses.entry(id).or_default();
+            entry.1 |= 1 << node_idx;
         }
         ast::CmpOp::Lt | ast::CmpOp::LtE if !is_left => {
-            uses.entry(id).or_default().upper_bound.insert(node_idx);
+            let entry = uses.entry(id).or_default();
+            entry.1 |= 1 << node_idx;
         }
         ast::CmpOp::Gt | ast::CmpOp::GtE if !is_left => {
-            uses.entry(id).or_default().lower_bound.insert(node_idx);
+            let entry = uses.entry(id).or_default();
+            entry.0 |= 1 << node_idx;
```

Comment:
now that you are using a bit-set, these will fail for `node_idx >= 32`

I recommend having a check at the top of `unneccessary_chained_comparison` for `bool_op.values.len() > 32`

I'm guessing we want to keep the HashMap implementation as a fallback for longer expressions